Code/GraphMol/NewCIPLabelling/Digraph.hpp

```diff
@@ -0,0 +1,258 @@
+#include "Edge.hpp"
+#include "TooManyNodesException.hpp"
+
+namespace RDKit {
+namespace NewCIPLabelling {
+
+template <typename A, typename B>
+class Node;
+
+namespace {
+
+/**
+ * Upper limit on the size of the digraph, stops out of memory error with a
+ * more graceful failure. 0=Infinite
+ */
+const int MAX_NODE_COUNT = 100000;
+
+/**
+ * Used for debugging only, 0=Infinite
+ */
+const int MAX_NODE_DIST = 0;
+}  // namespace
+
+template <typename A, typename B>
+class Digraph {
+ private:
+  BaseMol<A, B>* mol = nullptr;
+  Node<A, B>* root = nullptr;
+  Node<A, B>* tmproot = nullptr;
+  int numNodes = 0;
+  A rule6Ref = nullptr;
+
+  void addEdge(Node<A, B>* beg, B bond, Node<A, B>* end) {
+    auto e = new Edge<A, B>(beg, end, bond);
+    beg->add(e);
+    end->add(e);
+  }
+
+ public:
+  Digraph() = default;
+
+  Digraph(BaseMol<A, B>* mol) : mol{mol} {}
+
+  Digraph(BaseMol<A, B>* mol, A atom) : mol{mol} { init(atom); }
+
+  ~Digraph() { delete getCurrRoot(); }
+
+  BaseMol<A, B>* getMol() const { return mol; };
+
+  Node<A, B>* getRoot() const { return root; };
+
+  Node<A, B>* getCurrRoot() const {
+    return tmproot == nullptr ? root : tmproot;
+  }
+
+  int getNumNodes() const { return numNodes; }
+
+  std::vector<Node<A, B>*> getNodes(A atom) const {
+    auto result = std::vector<Node<A, B>*>{};
+    auto queue = std::queue<Node<A, B>*>{};
+    queue.push(getCurrRoot());
+
+    while (!queue.empty()) {
+      auto node = queue.front();
+      queue.pop();
+
+      if (atom == node->getAtom()) {
+        result.push_back(node);
+      }
+      for (const auto& e : node->getEdges()) {
+        if (!e->isBeg(node)) {
+          continue;
+        }
+        queue.push(e->getEnd());
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Access the reference atom for Rule 6 (if one is set).
+   */
+  A getRule6Ref() const { return rule6Ref; }
+
+  /**
+   * Used exclusively for Rule 6, we set one atom as the reference.
+   * @param ref reference atom
+   */
+  void setRule6Ref(A ref) { rule6Ref = ref; }
+
+  Node<A, B>* init(A atom) {
+    auto visit = std::vector<char>(mol->getNumAtoms());
+    visit[mol->getAtomIdx(atom)] = 1;
+
+    auto atomic_num_num = mol->getAtomicNum(atom);
+    auto atomic_num_den = 1;
+    auto dist = 1;
+    auto flags = 0x0;
+
+    root = new Node<A, B>(this, std::move(visit), atom, atomic_num_num,
+                          atomic_num_den, dist, flags);
+
+    ++numNodes;
+
+    return root;
+  }
+
+  /**
+   * Sets the root node of this digraph by flipping all the
+   * 'up' edges to be 'down'.
+   *
+   * @param newroot the new root
+   */
+  void changeRoot(Node<A, B>* newroot) {
+    auto queue = std::queue<Node<A, B>*>();
+    queue.push(newroot);
+
+    auto toflip = std::vector<Edge<A, B>*>();
+    while (!queue.empty()) {
+      auto node = queue.front();
+      queue.pop();
+
+      for (const auto& e : node->getEdges()) {
+        if (e->isEnd(node)) {
+          toflip.push_back(e);
```

Comment:
Yeah, that is one of the things where I think we might make a small improvement: I think it could be flipped here if we also updated the atom that is put in the queue, or just reverse the order of these statements.

But I didn't want to do any experimenting until this had been thoroughly tested -- this seems like one small change that could cause a bug that takes hours of debugging to find.